### PR TITLE
Implement Student's T distribution

### PR DESCRIPTION
**Pull request by alicanb**
`test_studentT_sample_grad` fails, I can understand `df<=1e-2` cases, but I have no idea why it fails for the rest.

### CODE REVIEW

File: test/test_distributions.py [hunk 1]
```diff
@@ -616,6 +624,59 @@ def test_chi2_sample_grad(self):
...
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    def test_studentT_sample(self):
+        set_rng_seed(0)  # see Note [Randomized statistical tests]
+        for df in [0.1, 1.0, 5.0]:
+            self._check_sampler_sampler(StudentT(df),
+                                        scipy.stats.t(df),
+                                        'StudentT(df={})'.format(df))
+
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    def test_studentT_sample_grad(self):
```
Comment (fritzo) [inline test/test_distributions.py]: It's nice to have all these tests of `standard_gamma_grad()`.

File: torch/distributions/studentT.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        >>> m.sample()  # Student's t-distributed with degrees of freedom=2
+         0.1046
+        [torch.FloatTensor of size 1]
+
+    Args:
+        df (float or Tensor or Variable): degrees of freedom
+    """
+    has_rsample = True
+
+    def __init__(self, df):
```
Comment (fritzo) [inline torch/distributions/studentT.py]: Tensorflow additionally has `loc=0`,`scale=1` parameters, and I think these are quite useful in practice where Student's t is used for "Gaussians with outliers". I think it wouldn't be much work to add those parameters, just a couple affine transforms and a `log(scale)` term in the `log_prob` and `entropy`? What do you think?

File: test/test_distributions.py (hunk 1 above)
Comment (fritzo) [inline test/test_distributions.py]: I believe this test cannot pass because the reparameterized gradients are stochastic and are only correct in expectation. It is safe to remove these tests.

The reason the gradients are stochastic is that our sampler contains a nuisance variable (projects the pair `X`, `Z` down to a single number). To compute zero-variance reparameterized gradients we would need to explicitly compute the cdf gradient. That's a lot of work (see implementations of `standard_gamma_grad()` and `dirichlet_grad()`). Actually my original implementation of `Beta` and `Dirichlet` had stochastic gradients for the same reason: I simply relied on `Gamma.rsample()` but that contains the nuisance variable `gammas.sum(-1)` which is projected out during normalization. I really wanted zero-variance gradients for `Dirichlet` (so we could implement e.g. LDA) so I put a bunch of work into implementing `dirichlet_grad()`.

Comment (fritzo) [reply test/test_distributions.py]: To see this, you should be able to plot reparameterized gradients as a function of df, using quantities computed in this test:
```py
dfs = Variable(10 ** torch.arange(-2, 4, 0.01))
actual_grad = ...
expected_grad = ...
pyplot.plot(df, expected_grad, 'k--', label='truth')
pyplot.scatter(df, actual_grad, label='approx')
pyplot.xscale('log')
pyplot.yscale('log')
pyplot.legend()
```

Comment (fritzo) [reply test/test_distributions.py]: Also see example plots (click on Detals) in the Dirichlet PR description: https://github.com/probtorch/pytorch/pull/28

File: torch/distributions/studentT.py (hunk 2 above)
Comment (alicanb) [reply torch/distributions/studentT.py]: Sure I can add it. There are a number of distributions in addition to student's T to which we can add `loc and scale`. Should I make a pass over and add them to add applicable ones?

Comment (fritzo) [reply torch/distributions/studentT.py]: Only if it helps improve compatibility with Tensorflow Distributions. I'm not a fan of scipy's design choice to add a `loc` parameter to nearly every distribution. I think what we would be more useful would be an `AffineBijector` and `TransformedDistribution` (working on those sections of the design doc right now, in case you want to review).

**Comment by alicanb** [comment]
Yes, I want to be involved with that, will take a look soon.

**Comment by alicanb** [comment]
`test_studentT_sample` fails when `loc!=0` and `scale!=1`  though...

**Comment by fritzo** [comment]
> `test_studentT_sample` fails if it's not `loc=0` and `scale=1` though...

Hmm that's weird. It looks like there are some real numerical issues with `df<=0.1` when using `FloatTensor`s, but that this is resolved by using `DoubleTensor`s (see this [notebook](https://github.com/fritzo/notebooks/blob/master/debug-student-t.ipynb)), however I was still unable to get tests to pass by wrapping parameters in `DoubleTensor`:
```diff
  for df, loc, scale in product(...):
-     self._check_sampler_sampler(StudentT(df, loc, scale),
+     self._check_sampler_sampler(StudentT(df=torch.DoubleTensor([df]),
+                                          loc=torch.DoubleTensor([loc]),
+                                          scale=torch.DoubleTensor([scale])),
                                  scipy.stats.t(df=df, loc=loc, scale=scale),
                                  'StudentT(df={}, loc={}, scale={})'.format(df, loc, scale))
```
